Namespaces and topics specified as parameters.

MFernandezCarmona opened the pull request:
Also some other minor changes:
people_tracker_filter/src/filter.cpp: changed method to acknowledge map reception. 
upper_body_detector: merged with ENRICHME branch, solves most of the memory problems.

MFernandezCarmona committed:
Merged with ENRICHME branch. Parametrized topics/frame_ids

marc-hanheide commented:
can you check why this failed? See https://lcas.lincoln.ac.uk/buildfarm/job/Kpr__strands_perception_people__ubuntu_xenial_amd64/2/console

This is the problem:

```
15:11:34 CMake Error at CMakeLists.txt:2 (project):
15:11:34   The CMAKE_C_COMPILER:
15:11:34 
15:11:34     /usr/lib/ccache/cc
15:11:34 
15:11:34   is not a full path to an existing compiler tool.
15:11:34 
15:11:34   Tell CMake where to find the compiler by setting either the environment
15:11:34   variable "CC" or the CMake cache entry CMAKE_C_COMPILER to the full path to
15:11:34   the compiler, or to the compiler name if it is in the PATH.
15:11:34 
15:11:34 
15:11:34 CMake Error at CMakeLists.txt:2 (project):
15:11:34   The CMAKE_CXX_COMPILER:
15:11:34 
15:11:34     /usr/lib/ccache/c++
15:11:34 
15:11:34   is not a full path to an existing compiler tool.
15:11:34 
15:11:34   Tell CMake where to find the compiler by setting either the environment
15:11:34   variable "CXX" or the CMake cache entry CMAKE_CXX_COMPILER to the full path
15:11:34   to the compiler, or to the compiler name if it is in the PATH.
```

I recall a similar problem where the ROS dependencies have changes and one now needs to explicitly depend on roscpp or similar, I seem to recall.

lucasb-eyer commented on upper_body_detector/src/KConnectedComponentLabeler.cpp:
```diff
@@ -290,17 +296,22 @@ void KConnectedComponentLabeler::Process()
 
-		delete trueLabelArray; trueLabelArray = NULL;
+		//delete trueLabelArray; trueLabelArray = NULL;
+                trueLabelArray.clear();
```
Not needed, but doesn't hurt either.

lucasb-eyer commented on upper_body_detector/src/KConnectedComponentLabeler.cpp:
```diff
@@ -290,17 +296,22 @@ void KConnectedComponentLabeler::Process()
+	//lookUpTable = NULL;
+
+	//delete []regionArea;
+	//regionArea = NULL;
```
You can remove the comments, git keeps history :smile:

lucasb-eyer commented on upper_body_detector/src/KConnectedComponentLabeler.cpp:
```diff
@@ -290,17 +296,22 @@ void KConnectedComponentLabeler::Process()
-	delete []regionArea;
-	regionArea = NULL;
+	//delete []regionLabel;
+	//regionLabel = NULL;
```
same here

lucasb-eyer commented on upper_body_detector/src/KConnectedComponentLabeler.cpp:
```diff
@@ -290,17 +296,22 @@ void KConnectedComponentLabeler::Process()
-	regionLabel = NULL;
+        lookUpTable.clear();
+        regionArea.clear();
+        regionLabel.clear();
```
No need for `.clear()` as they get deleted when going out of scope here.

lucasb-eyer commented on upper_body_detector/src/detector.cpp:
```diff
@@ -124,12 +124,22 @@ void Detector::ComputeFreespace(const Camera& camera,
+                }
+                else
+                {
+                   ROS_INFO("upper_body_detector: compute freespace - pos_x:%i/%i, pos_z:%i/%i - out of range!", pos_x,x_bins,pos_z,z_bins);
```
Oh, that happened sometimes? Good catch, if so!

lucasb-eyer commented on upper_body_detector/src/detector.cpp:
```diff
@@ -144,9 +154,15 @@ void Detector::ComputeFreespace(const Camera& camera,
+        {
+	    //ROS_INFO("occ_map_binary -> x:%d, y:%i",x,y);
+            
+            occ_map_binary(x,y) = (occ_map(x,y) < Globals::freespace_threshold) ? 0 : 1;
```
It would be nice if you could rename `y` to `z`, as that's consistent with nomenclature elsewhere in the detector.

lucasb-eyer commented on upper_body_detector/src/main.cpp:
```diff
@@ -198,9 +281,34 @@ void callback(const ImageConstPtr &depth, const ImageConstPtr &color,const Groun
-        for (int c = 0;c < 640;c++) {
-            matrix_depth(c, r) = img_depth_.at<float>(r,c);
+    for (int r = 0;r < info->height;r++){
+        for (int c = 0;c < info->width;c++) {
```
Yes, finally, thank you :smile:

lucasb-eyer commented on upper_body_detector/src/main.cpp:
```diff
@@ -198,9 +281,34 @@ void callback(const ImageConstPtr &depth, const ImageConstPtr &color,const Groun
+            else
+            {
+               newVal = 0;
+            }
```
This block of code has several problems:

1) the second check is checking for NaNs, right? In the original code, NaNs were passed through, but with your change, they are converted to 0s. I think NaNs should still be passed through, as they signify something completely different from depth 0.

2) Interpreting the pixel type using DEPTH_SCALE sounds scary for me, imagine someone changes DEPTH_SCALE to 0.999 and everything breaks, no way they will understand why.

3) I'm not sure what value is used for NaN in the `ushort`, i.e. original encoding? But it should be converted to `NaN`, too. I'm assuming it's `0` in the following, but please do double-check to make sure.

I suggest something like the following (untested):

```
    int inttype = img_depth_.type();
    uchar type = inttype & CV_MAT_DEPTH_MASK;
    if (type == CV_32F) {
        matrix_depth(c, r) = img_depth_.at<float>(r,c) / Globals::DEPTH_SCALE;
    } else if (type == CV_16U) {
        float raw_val = img_depth_.at<ushort>(r,c);
        if (raw_val == 0) { // Please double-check if 0 corresponds to NaN.
            matrix_depth(c, r) = nanf();
        } else {
            matrix_depth(c, r) = raw_val / Globals::DEPTH_SCALE;
        }
    } else {
        ROS_ERROR("Unknown depth-map type, no detections.");
    }
```

MFernandezCarmona replied:
Hi,
The real human perception expert in ENRICHME was @scosar. I will check this with him later. In the meantime your solution seems more robust.

lucasb-eyer commented on upper_body_detector/src/main.cpp:
```diff
@@ -217,7 +325,27 @@ void callback(const ImageConstPtr &depth, const ImageConstPtr &color,const Groun
+    }
+    cv::cvtColor(image, image, CV_HSV2RGB);
+    cv_color_ptr = cv_bridge::toCvCopy(color);
+    cv_color_ptr->image = image;
```
This might have worked by chance, but this really doesn't look like correct code. Why not just `cv_color_ptr = cv_bridge::toCvCopy(image);` (note: `image`, not `color`.)

lucasb-eyer commented on upper_body_detector/include/detector.h:
```diff
@@ -15,7 +15,7 @@ class Detector
     Detector();
 
-    bool visualize_roi;
+    bool visualize_roi=false;
```
This effectively makes it C++11, which would be nice to avoid. It's not crucial, but it's a simple thing here.

lucasb-eyer commented on upper_body_detector/config/upper_body_detector.yaml:
```diff
@@ -48,12 +48,17 @@ upper_body_detector:
     # scanning corridor, which means for each ROI we scan it with the depth template starting from min_height (in meters) for up to max_height
     max_height: 2.1
-    min_height: 1.0
+    min_height: 0.85
```
I'm not sure it's a good idea to change these in the default file.
Knowing Dennis, he probably spent quite some time tuning them for his scenarios, and you tuned them for your scenarios.

I'd be very happy if you would leave the original parameters, make a new copy with your parameters (like you did with the `r2`), and then in the README quickly describe how to use them (and what scenario your parameters were tuned for).

lucasb-eyer commented on ground_plane_estimation/src/fixed_gp.cpp:
```diff
@@ -31,8 +32,8 @@ void callback(const sensor_msgs::JointState::ConstPtr &msg) {
-    gp.header.frame_id = "/head_xtion_depth_optical_frame";
-    gp.header.stamp = msg->header.stamp;
+    gp.header.frame_id = _frame_id;
+    gp.header.stamp = ros::Time::now();
```
Why the change? I think we've been careful to "pull through" the timestamps where needed, for example, the depth map's stamp gets pulled through the detector and tracker. So you need a good argument for the change here.

lucasb-eyer reviewed (changes requested):
Overall, I think these are great changes that make the detector much better. I commend you for going through that code and thanks a lot for contributing back :+1: :smile:

I left a few comments inline I'd like you to address, after that I think we can merge. From the diff, I don't see how it can cause the build failure (it can't find a compiler).

One more thing is, if I understand correctly, this could also make it work with Kinect v2, right? It would be great if you could add a section to the README with describing this and how to use it (basically just using the new config, but that's way not obvious for users who don't dig through the code like you did).

MFernandezCarmona committed:
Undone the timestamp change

Hi,
It was a careless change. I originally thought that the timestamp should reflect time creation of the data, but it also makes sense the way you made it plus it's better not to change the meaning of a field now.

MFernandezCarmona committed:
Reverted changes on original config.

MFernandezCarmona committed:
Update detector.h

MFernandezCarmona committed:
Default initialization on constructor

MFernandezCarmona committed:
Update KConnectedComponentLabeler.cpp

MFernandezCarmona committed:
Update KConnectedComponentLabeler.cpp

MFernandezCarmona committed:
Update detector.cpp

MFernandezCarmona committed:
Suggested change in data interpretation. 

Still pending to check that 0 corresponds to NaN in ushort

MFernandezCarmona committed:
Bug corrected.

MFernandezCarmona commented:
Hi,
Our fork was really maintained by @scosar. I will talk with him about your comments and corrections. I did use this with Kinect2, I will put more detailled description and probably an extra launcher later.
Cheers

marc-hanheide commented:
retest this please

marc-hanheide commented:
@MFernandezCarmona I managed to fix the issues around `ccache` on the integration server, but there as indeed relevant issues still to be resolve, see https://lcas.lincoln.ac.uk/buildfarm/job/Kpr__strands_perception_people__ubuntu_xenial_amd64/12/console

```
18:32:10 [ 40%] Building CXX object CMakeFiles/upper_body_detector.dir/src/main.cpp.o
18:32:17 /tmp/catkin_workspace/src/strands_perception_people/upper_body_detector/src/main.cpp: In function ‘void callback(const ImageConstPtr&, const ImageConstPtr&, const ConstPtr&, const CameraInfoConstPtr&)’:
18:32:17 /tmp/catkin_workspace/src/strands_perception_people/upper_body_detector/src/main.cpp:294:33: error: too few arguments to function ‘float nanf(const char*)’
18:32:17        matrix_depth(c, r) = nanf();
18:32:17                                  ^
18:32:17 In file included from /usr/include/math.h:104:0,
18:32:17                  from /usr/include/c++/5/cmath:44,
18:32:17                  from /opt/ros/kinetic/include/ros/time.h:55,
18:32:17                  from /opt/ros/kinetic/include/ros/ros.h:38,
18:32:17                  from /tmp/catkin_workspace/src/strands_perception_people/upper_body_detector/src/main.cpp:2:
18:32:17 /usr/include/x86_64-linux-gnu/bits/mathcalls.h:228:14: note: declared here
18:32:17  __MATHCALLX (nan,, (const char *__tagb), (__const__));

```

lucasb-eyer commented:
Thanks for adressing all these, and thanks again for merging back upstream!

The build failure is because I forgot that `nanf()` needs to be called as `nanf("")` instead, that should fix it.

MFernandezCarmona committed:
commented duplicate code
